Deploy any version

mbarnes opened the pull request:
Following up on https://github.com/projectatomic/rpm-ostree/issues/172 and https://github.com/GNOME/ostree/pull/147.

I'm still working on test cases for this, but I wanted to get the code changes posted ASAP.

This enhances the `rpm-ostree rebase` command to allow rebasing to any version.

Syntax changes are:
```
rpm-ostree rebase [-V / --to-version=VERSION] [-R / --to-revision=CHECKSUM] [REFSPEC]
```
Note: `REFSPEC` is now optional, but only if `--to-version` or `--to-revision` is specified.  Defaults to the booted deployment's origin refspec.

I know we talked about putting version lookup directly in `libostree` and I still want to, but I'm shying away from it for the moment because I think it needs more thinking through.  The way I'm solving version lookups here is really slow and clunky and I'd like to come up with a better way before messing with `libostree`.  (Something analogous to git tags feels more like the Right Way to me.  Available tags could be included in summary files.)

petervo commented:
If I'm understanding this correctly, to ensure you only upgrade to a revision that you've seen the diff/checksum for already like cockpit wants to do, you would need to always run a rebase not an upgrade. Is that right?

As long as rpm-ostree is used, running an upgrade later will clear the flag so everything will work as expected. i imagine it could lead to some confusion if someone falls back to using ostree directly later if they "upgraded" in this way when underneath it was really a revision pegged rebase. That can probably be fixed with documentation.

Another option might be to add the to-revision argument to upgrade, that would set the override-commit flag run the update and then clear it. Or if that is too much flexibilty, allowing people to do things that really are not upgrades, we could have a --current-head-only option that sets override-commit to the current head and the clears it when the upgrade is done.

mbarnes commented:
> If I'm understanding this correctly, to ensure you only upgrade to a revision that you've seen the diff/checksum for already like cockpit wants to do, you would need to always run a rebase not an upgrade. Is that right?

I think we could easily tweak the upgrade and rollback methods to ensure you're moving to the commit you think you are without resorting to this feature.  That's probably better left for Friday's design meeting.

cgwalters commented:
One issue to note is that by default `ostree_sysroot_cleanup` as used by the various tools will prune the repo with `_REFS_ONLY` and `depth = 0`, so the intermediate history will be deleted.  (

The problem with just changing it to `depth = -1` is that we could end up hanging on to *full* trees with normal upgrades which would take up a lot of space.   I'll file a bug on this.

cgwalters commented:
https://bugzilla.gnome.org/show_bug.cgi?id=756639

cgwalters commented:
One thing I'm thinking about here is the `OSTREE_SYSROOT_UPGRADER_PULL_FLAGS_ALLOW_OLDER` flag - the `rebase` option always provides this because the assumption is there won't be a chronological relationship between different branches.  (Perhaps we should make this an option still, as if you're going from a "prod" branch to a "dev" branch, the latter should always be newer)

Anyways what I'm thinking about here is that if use of "upgrade to version" becomes common, then we effectively lose the protection against replay attacks that the timestamp check provides.  This might argue then for an `UpgradeTo` DBus method as opposed to using rebase?

cgwalters commented:
Above aside, code is looking pretty good to me!

mbarnes commented:
> Anyways what I'm thinking about here is that if use of "upgrade to version" becomes common, then we effectively lose the protection against replay attacks that the timestamp check provides. This might argue then for an UpgradeTo DBus method as opposed to using rebase?

I extended `Rebase` instead of `Upgrade` because it seemed like the better verb.  Like I said elsewhere, calling `Upgrade` to downgrade seems weird to me.

But another option might be to extend `Upgrade` for *upgrading* to a newer-but-not-latest version and introduce `Downgrade` for downgrading -- each with appropriate checks -- and leave `Rebase` alone.

mbarnes commented:
Added progress to version lookup.

As for GPG verification, version lookup is just using `ostree_repo_pull_with_options()` and since https://github.com/GNOME/ostree/commit/ed861609756202cc0a23a85a5fab3d85c880e1c2, recursive commit-only pulls pass through `scan_commit_object()` which does the GPG verification.  So I expect we should be good there, but I'll add a test case for it.

mbarnes commented:
Seems clear from our usability meeting with @stefwalter and @petervo that we'll want to split this feature out from the `Rebase` method.  I'll set test cases aside for now and start reworking this.

mbarnes commented:
Complete rewrite here.  Still working on test cases.

I abandoned trying to bolt this on to the `rebase` command and instead created a new `deploy` command and associated D-Bus methods:

```
Deploy(revision 's', options 'a{sv}') -> transaction_address 's'

DownloadDeployRpmDiff(revision 's', packages 'as') -> transaction_address 's'

GetCachedDeployRpmDiff(revision 's', packages 'as') -> (result 'a(sua{sv}), details 'a{sv}')
```

And the new command syntax is:
```
rpm-ostree deploy [--preview] [--reboot] REVISION
```

In all cases, `revision` must be either a full SHA256 checksum or a version value.  For now, the daemon handles the version-to-checksum lookup.

The `deploy --preview` option is similar to `upgrade --check-diff` in that it only prints a package diff.

jlebon commented:
bot, retest this please

mbarnes commented:
Rebased on master and added some new commits and fixups, which GitHub seems to have shuffled again.

Some refactoring at the CLI layer (https://github.com/mbarnes/rpm-ostree/commit/99ecf6601fea7d5f2c2e300b9765734e9c35308d) makes way for tri-state exit codes (https://github.com/mbarnes/rpm-ostree/commit/524edc622dd14f258a829910541550c85d96eb39).

Also, the last commit implements the `upgrade` option changes I had suggested, but wasn't clear if @cgwalters was in favor of it.  So I can throw it out if there's objections.

cgwalters commented:
One minor comment, otherwise it looks pretty nice to me!  Thanks a lot for the well commented patch series.

cgwalters commented:
i.e. :+1: with or without above comment

mbarnes commented:
Implemented the prefixes on the revision argument and finally got tests done.

cgwalters commented:
LGTM!  :+1:

cgwalters commented:
This broke the build on our internal RHEL7:

```
src/daemon/rpmostreed-os.c: In function 'os_handle_get_cached_deploy_rpm_diff':
src/daemon/rpmostreed-os.c:1068:3: error: unknown type name 'GCancellable_autoptr'
   g_autoptr(GCancellable) cancellable = NULL;
```

Need another libglnx push?

mbarnes commented:
Should be fixed now.
